### PR TITLE
Add olm + catalogs to mgmt cluster

### PR DESCRIPTION
**Pull request by ecordell**
Testing:

```bash
./bin/hypershift install --hypershift-image quay.io/ecordell/hypershift:mgmtcatalogs14
./bin/hypershift create cluster --pull-secret=./build01config.json --aws-creds=<aws creds> --name="guest" --base-domain=devcluster.openshift.com --release-image=<image>
```
- `./build01config.json` should be a pull secret for the build01 ci cluster
- `<image>` should be a release image from cluster-bot built with `build openshift/operator-framework-olm#69,operator-framework/operator-marketplace#396`
- the `pull-secret` for the ci-cluster needs to be added to the default serviceaccount both in the `hypershift` namespace and in the `clusters-guest` namespace so that `imagelookup` pods succeed

Results:

Screen Shot 2021-05-13 at 11 07 49 AM

- default catalogs are deployments in the host cluster
- default catalogs auto-update via imagestreams with scheduled import (instead of relying on the host's catalog-operator to update)
- actual CatalogSource objects for the defaults are in the normal place in the guest (`openshift-marketplace`) but instead of referencing the catalog images directly, they point to the service address of the catalogs in the host. 
- additional CatalogSources can be added to the guest
- catalog-operator and packageserver need to talk to services in the guest and the host (for user-provided and default catalogs), so they both have the vpn sidecar and a dnsmasq sidecar (that can check both guest and host dns to resolve catalog endpoints)

**TODO**

Guest nodes lack credentials for redhat registries - this means many operators fail to install, for example:

```
Failed to pull image "registry.redhat.io/openshift4/ose-clusterresourceoverride-operator-bundle@sha256:081a1e247e2d4be1b502b7e336d73ae03d0203bd1ad8d03ce2d7bf61ae969ebe": rpc error: code = Unknown desc = unable to retrieve auth token: invalid username/password: unauthorized: Please login to the Red Hat Registry using your Customer Portal credentials. Further instructions can be found here: https://access.redhat.com/RegistryAuthentication
```

**Resource Usage**

Host resources used with 1 custom catalog (on guest), 4 default catalogs, 1 operator installed, otherwise at rest.
```
catalog-operator 364.9 MiB	0.005 cores
olm-operator     121.2 MiB	0.008 cores
packageserver    91.7 MiB	0.020 cores
packageserver    86.2 MiB	0.020 cores
---------------------------------------
all operators    664 MiB 0.053 cores


certified-operators 135.4 MiB	0.016 cores
community-operators 181.8 MiB	0.019 cores
redhat-marketplace  110.4 MiB	0.012 cores
redhat-operators    191.6 MiB	0.014 cores
-------------------------------------------
all catalogs        619   MiB   0.061 cores

total 1283 MiB, 0.114 cores
```

Potential savings:
 - packageserver should only have 1 replica (- 90MiB)
 - catalogs should stream their contents instead of building the entire response in memory (-500MiB)
 - catalog-operator package cache should (configurably) flush to disk instead of staying in memory (-200MiB)
 - improve at-rest client-go caching (?? MiB)
 
target total: ~400MiB

### CODE REVIEW

**Comment by ironcladlou** [comment]
Will need reconciled with @csrwng's work in https://github.com/openshift/hypershift/pull/200 which we expect to merge next week, we're moving away from the template system

File: control-plane-operator/controllers/hostedcontrolplane/assets/olm/0000_50_olm_00-namespace.yaml.notneeded [hunk 1]
```diff
@@ -0,0 +1,26 @@
+apiVersion: v1
```
Comment (enxebre) [inline control-plane-operator/controllers/hostedcontrolplane/assets/olm/0000_50_olm_00-namespace.yaml.notneeded]: are the .notneeded files needed?

File: control-plane-operator/controllers/hostedcontrolplane/assets/olm/0000_50_olm_00-catalogsources.crd.yaml.guest [hunk 2]
```diff
@@ -0,0 +1,169 @@
+apiVersion: apiextensions.k8s.io/v1
```
Comment (enxebre) [inline control-plane-operator/controllers/hostedcontrolplane/assets/olm/0000_50_olm_00-catalogsources.crd.yaml.guest]: can we stick to the .yaml extension and may create an inner folder for guest manifests?

File: control-plane-operator/controllers/hostedcontrolplane/render/manifests.go [hunk 3]
```diff
@@ -358,6 +359,21 @@ func (c *clusterManifestContext) ignitionConfigs() {
...
+func (c *clusterManifestContext) operatorLifecycleManager() {
+	c.addManifestFiles(
+		"olm/catalog-metrics-service.yaml",
+		"olm/olm-metrics-service.yaml",
+		"olm/0000_50_olm_07-olm-operator.deployment.yaml",
+		"olm/0000_50_olm_08-catalog-operator.deployment.yaml",
+		"olm/packageserver.deployment.yaml",
+	)
+	/*
+		c.addUserManifestFiles(
```
Comment (enxebre) [inline control-plane-operator/controllers/hostedcontrolplane/render/manifests.go]: please drop comment.

File: control-plane-operator/controllers/hostedcontrolplane/assets/assets.go [hunk 4]
```diff
@@ -22,6 +22,7 @@ import "embed"
 //go:embed registry/*
 //go:embed roks-metrics/*
 //go:embed user-manifests-bootstrapper/*
+//go:embed olm/*
 var content embed.FS
 
 func AssetDir(name string) ([]string, error) {
```
Comment (enxebre) [inline control-plane-operator/controllers/hostedcontrolplane/assets/assets.go:29]: Please let's use imperative mood in the commit subject.

File: control-plane-operator/controllers/hostedcontrolplane/assets/olm/0000_50_olm_00-namespace.yaml.notneeded [hunk 5]
```diff
@@ -1,26 +0,0 @@
-apiVersion: v1
```
Comment (enxebre) [inline control-plane-operator/controllers/hostedcontrolplane/assets/olm/0000_50_olm_00-namespace.yaml.notneeded]: is this commit undoing changes introduced by a previous commit? Please let's consolidate those and squash them in one with final outcome.

File: .gitignore [hunk 6]
```diff
@@ -1,4 +1,4 @@
-
+hypershift
 # Binaries for programs and plugins
```
Comment (enxebre) [inline .gitignore:2]: did this sneaked in or is intentional?

File: control-plane-operator/controllers/hostedcontrolplane/render/manifests.go [hunk 7]
```diff
@@ -379,7 +382,9 @@ func (c *clusterManifestContext) operatorLifecycleManager() {
 		"olm/operators-crd-guest.yaml",
 		"olm/subscriptions-crd-guest.yaml",
 		"olm/openshift-operators-namespace-guest.yaml",
+		"olm/openshift-operator-lifecycle-manager-namespace-guest.yaml",
 		"olm/global-operatorgroup-default-guest.yaml",
+		//"olm/prometheus-rule-guest.yaml",
```
Comment (enxebre) [inline control-plane-operator/controllers/hostedcontrolplane/render/manifests.go:409]: please drop comment.

File: control-plane-operator/controllers/hostedcontrolplane/assets/olm/catalog-certified-operators.yaml [hunk 8]
```diff
@@ -1,15 +0,0 @@
-apiVersion: "operators.coreos.com/v1alpha1"
```
Comment (enxebre) [inline control-plane-operator/controllers/hostedcontrolplane/assets/olm/catalog-certified-operators.yaml]: This seems to be undoing a change from a previous commit? Let's consolidate it in one atomic change with the expected result.

Comment (enxebre) [inline control-plane-operator/controllers/hostedcontrolplane/assets/olm/catalog-certified-operators.yaml]: Please let's use imperative mood in the commit subject (action) and elaborate the motivation in the commit description, i.e git ci -m"Imperative mood action" -m"Why"

**Comment by ironcladlou** [comment]
I want to discuss how to reconcile this with @csrwng's refactor away from templates in this operator; merging ~35 more right now seems like it could be fatal to the effort. I propose Cesar's refactor merges first and then all new contributions to the operator (including this one) follow the new template-free pattern.

/hold

**Comment by ironcladlou** [comment]
I also want to revisit the DNS discussion to understand what we can do with CoreDNS to avoid managing dnsmasq for any reason whatsoever.

**Comment by csrwng** [comment]
/close
replaced by https://github.com/openshift/hypershift/pull/265

**Pull request closed without merging**